src/iterator.c

```diff
@@ -1482,6 +1482,427 @@ int git_iterator_for_workdir_ext(
+} workdirfilelist_iterator;
+
+static int workdirfilelist_stat(struct stat *st, git_iterator *self, const char *path)
+{
+	workdirfilelist_iterator *wdfi = (workdirfilelist_iterator *)self;
+	git_buf buf_fullpath = GIT_BUF_INIT;
+
+	git_buf_joinpath(&buf_fullpath, wdfi->buf_workdir.ptr, path);
+	return (p_lstat(buf_fullpath.ptr, st));
+}
+
+/* The "end" string should be included in the enumeration,
+ * so we set the "pos_filelist_end" value just past its
+ * position (in STL style).
+ */
+static int workdirfilelist_iterator__at_end(git_iterator *self)
+{
+	workdirfilelist_iterator *wdfi = (workdirfilelist_iterator *)self;
+	return (wdfi->pos_filelist_current >= wdfi->pos_filelist_end);
+}
+
+static int workdirfilelist_iterator__current(
+	const git_index_entry **entry, git_iterator *self)
+{
+	workdirfilelist_iterator *wdfi = (workdirfilelist_iterator *)self;
+
+	iterator__clear_entry(entry);
+
+	if (workdirfilelist_iterator__at_end(self))
+		return GIT_ITEROVER;
+
+	if (entry)
+		*entry = &wdfi->entry;
+
+	wdfi->base.flags |= GIT_ITERATOR_FIRST_ACCESS;
+	return 0;
+}
+
+static int workdirfilelist_iterator__advance(
+	const git_index_entry **entry, git_iterator *self)
+{
+	workdirfilelist_iterator *wdfi = (workdirfilelist_iterator *)self;
+
+	if (!iterator__has_been_accessed(wdfi))
+		return workdirfilelist_iterator__current(entry, self);
+
+	wdfi->pos_filelist_current++;
+
+	while (!workdirfilelist_iterator__at_end(self)) {
+		struct stat st;
+		const char *sz = git_vector_get(&wdfi->filelist, wdfi->pos_filelist_current);
+		if (workdirfilelist_stat(&st, self, sz) == 0) {
+			git_buf_sets(&wdfi->buf_entry_path, sz);
+			wdfi->entry.path = wdfi->buf_entry_path.ptr;
+			git_index_entry__init_from_stat(&wdfi->entry, &st, true);
+			break;
+		}
+		wdfi->pos_filelist_current++;
+	}
+	return workdirfilelist_iterator__current(entry, self);
+}
+
+static int workdirfilelist_iterator__advance_into(
+	const git_index_entry **entry, git_iterator *self)
+{
+	/* not needed since we don't present a diving/treewalk concept. */
+	GIT_UNUSED(entry); GIT_UNUSED(self);
+	return GIT_ENOTFOUND;
+}
+
+static int workdirfilelist_iterator__seek(
+	git_iterator *self, const char *prefix)
+{
+	GIT_UNUSED(self); GIT_UNUSED(prefix);
+	return -1;
+}
+
+static int workdirfilelist_iterator__reset(
+	git_iterator *self, const char *start, const char *end)
+{
+	workdirfilelist_iterator *wdfi = (workdirfilelist_iterator *)self;
+
+	if (iterator__reset_range(self, start, end) < 0)
+		return -1;
+
+	/* We lookup the starting and ending paths in the filelist.
+	 * This gives us the bounds on set
+	 * of possible paths we will return during the iteration.
+	 *
+	 * For the "end" string, we want it to be included in the
+	 * set of paths enumerated (unlike the STL).  We set the
+	 * "pos_filelist_end" to be just past it (like the STL).
+	 */
+
+	wdfi->pos_filelist_start = 0;
+	if (wdfi->base.start)
+		git_vector_bsearch(&wdfi->pos_filelist_start, &wdfi->filelist, wdfi->base.start);
```

Comment:
Hm, I realised you're probably relying on the start position to remain untouched if we do not find it. I suppose that's fine since we try to stick to not changing the output parameter if we fail the call.